### PR TITLE
fixes #20244; fixes castSizes warnings

### LINKED ISSUE #20244: strutils.nim(954, 13) Warning: target type is larger than source type [CastSizes]
### What happened?

I am talking about examples from https://github.com/khchen/winim/tree/master/examples/clr

when I compile winim-master\examples\clr\simple_gui.nim with nim 1.7.1, the message says
```bash
E:\msys64\home\USER\_nim\nim\lib\pure\strutils.nim(954, 13) Warning: target type is larger than source type [CastSizes]
```

### Nim Version

latest cloned-and-compiled nim 1.7.1

```
$ nim -v
Nim Compiler Version 1.7.1 [Windows: amd64]
Compiled at 2022-08-16
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: https://github.com/nim-lang/Nim/commit/9f408ea9430b104eb6c14a84d7a1650bf115b6a0
active boot switches: -d:release
```

### Current Standard Output Logs

_No response_

### Expected Standard Output Logs

```shell
there is no such information

E:\msys64\home\USER\_nim\nim\lib\pure\strutils.nim(954, 13) Warning: target type is larger than source type [CastSizes]
```

### Possible Solution

_No response_

### Additional Information

I don't know whether this is porblem in nim or winim, but since the indicated file is nim's lib, I report it as a bug here.

### PR DESCRIPTION
**Pull request by ringabout**
fixes #20244

### CODE REVIEW

File: compiler/ccgutils.nim [hunk 1]
```diff
@@ -53,7 +53,9 @@ proc hashString*(conf: ConfigRef; s: string): BiggestInt =
     a = a + (a shl 3)
     a = a xor (a shr 11)
     a = a + (a shl 15)
+    {.push warnings: off.}
     result = cast[Hash](a)
```
Comment (arnetheduck) [inline compiler/ccgutils.nim]: cast will generally fill the extra bytes with random garbage when creating a bigger type - this doesn't smell safe in a hash function

Comment (arnetheduck) [reply compiler/ccgutils.nim]: ie, it doesn't smell safe anywhere really, except when you're about to overwrite the rest of the bytes with other data - the better fix here would actually be to not use a `cast` and instead use `BiggestInt(a)`

File: lib/std/syncio.nim [hunk 2]
```diff
@@ -376,7 +376,7 @@ when defined(nimdoc) or (defined(posix) and not defined(nimscript)) or defined(w
       flags = if inheritable: flags and not FD_CLOEXEC else: flags or FD_CLOEXEC
       result = c_fcntl(f, F_SETFD, flags) != -1
     else:
-      result = setHandleInformation(cast[IoHandle](f), HANDLE_FLAG_INHERIT,
+      result = setHandleInformation(cast[IoHandle](int(f)), HANDLE_FLAG_INHERIT,
```
Comment (arnetheduck) [inline lib/std/syncio.nim]: why is a conversion to int needed here? if f is smaller than iohandle, this would be a bug

Comment (ringabout) [reply lib/std/syncio.nim]: `f` is a `cint`, `IoHandle` is a `distinct pointer`.

**Comment by arnetheduck** [comment]
This is the second PR recently where the cast warning is suggested to be disabled instead of addressing the problem it warns about - I think the warning text needs to be updated to point out the garbage data problem with size extension which experience shows is not obvious / well known

File: lib/std/syncio.nim (hunk 2 above)
Comment (arnetheduck) [reply lib/std/syncio.nim]: that sounds fishy - how can that work even? windows functions like createfile return a `Handle`, not a cint afair - either that handle is  being truncated or this function is being called with garbage from different subsystems (stdio emulation vs native windows handles)

Comment (ringabout) [reply lib/std/syncio.nim]:
```nim
type
  FileHandle* = cint ## type that represents an OS file handle; this is
                      ## useful for low-level file access
```

https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/fdopen-wfdopen?view=msvc-170

The file descriptor of the open file is `cint` on Windows.

Comment (arnetheduck) [reply lib/std/syncio.nim]: no, that's the C emulation layer - files are Handle on windows: 

https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea

To us other "native" functions, c file numbers must be converted first:
https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/get-osfhandle?source=recommendations&view=msvc-170

Comment (ringabout) [reply lib/std/syncio.nim]: https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/get-osfhandle?view=msvc-170


```nim
proc getOsFileHandle*(f: File): FileHandle =
  ## Returns the OS file handle of the file `f`. This is only useful for
  ## platform specific programming.
  when defined(windows):
    result = FileHandle getOsfhandle(cint getFileHandle(f))
  else:
    result = c_fileno(f)
```

It is truncated here.

Comment (arnetheduck) [reply lib/std/syncio.nim]: :scream:

Comment (ringabout) [inline lib/std/syncio.nim]:
```suggestion
      result = setHandleInformation(cast[IoHandle](f), HANDLE_FLAG_INHERIT,
```

Anyway, it is not the scope of this PR which intends to fix #20244.

**Comment by ringabout** [comment]
> This is the second PR recently where the cast warning is suggested to be disabled instead of addressing the problem it warns about

I did look at the original PR and knew the effect of garbage bits. I used warnings for the compiler because I mixed it with `when defined`.

File: lib/std/syncio.nim (hunk 2 above)
Comment (ringabout) [reply lib/std/syncio.nim]: Well, we should deprecate it and fixes its use case in the stdlib.

Comment (ringabout) [reply lib/std/syncio.nim]: A bit breaking change => https://github.com/nim-lang/Nim/pull/21106 perhaps we need to add a new proc

**Review by arnetheduck — APPROVED** [review]

**Pull request merged**